src/background_iterator.rs

```diff
@@ -0,0 +1,54 @@
+use std;
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread;
+
+/// Execute an iterator on a worker thread, which can work ahead a configurable number of items.
+pub struct BackgroundIterator<T> {
+    rx: Receiver<Option<T>>,
+    done: bool,
+}
+
+impl<T: Send> Iterator for BackgroundIterator<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<T> {
+        if self.done {
+            return None;
+        }
+
+        match self.rx.recv() {
+            Ok(Some(v)) => Some(v),
+            Ok(None) => {
+                self.done = true;
+                None
+            }
+
+            // if the producer thread dies, the iterator
+            // silently exits, without an error.
+            Err(_) => None,
```

Comment:
Should this maybe return the error instead, in case for example you had multiple senders or something like that and one of them died...

Also what happens if the sending thread panics?  It's not clear to me how error propagation works here, though of course that might just because I'm not familiar with threading in rust.

Also should this not set `self.done`?

Reply:
good call.  we now keep the join handle & if we get an error on `recv` we check it for a panic & propagate to the reading thread.  In the original code we could swallow a panic and potentially proceed with partial results.